### ISSUE TITLE
Retrying HTTP requests help

### ISSUE THREAD

**Issue by moose56** [caolan/highland#641]
Hi, 

building on #640 where I download a series of gziped files and stream them into a target file on disk, I want to add in some retry functionality into the requests using a module like [retry-request](https://github.com/stephenplusplus/retry-request).

```js
const zlib = require('zlib');
const fs = require('fs');
const request = require('retry-request');
const _ = require('highland');

function download (urls, path) {
  const tmpPath = path + '-tmp';
  const unzip = zlib.createGunzip();
  const output = fs.createWriteStream(tmpPath);

  const getUrl = url => {
    return _((push, next) => {
      const options = {
        url,
        retries: 10
      };
      request(options)
        .on('response', resp => {
          if (resp.statusCode === 200) {
            push(null, _(resp));
            push(null, _.nil);
          } else {
            push(new Error('Something went wrong'));
            push(null, _.nil);
          }
        });
    });
  };

  let error;

  return new Promise((resolve, reject) => {
    _(urls)
      .flatMap(getUrl)
      .parallel(5)
      .through(unzip)
      .stopOnError(err => error = err)
      .pipe(output)
      .on('finish', () => {
        console.log('finish');
        if (error) return reject(error);

        fs.rename(tmpPath, path, (err) => {
          if (err) { return reject(err); }
          return resolve(path);
        });
      })
      .on('error', reject);
  });
}
```
What happens when I run this is the first 5 URLs are downloaded OK and then execution just stops. There appear to be no errors so I am not sure what is happening. 

Anybody able to shed any light on this behaviour and why it might not be working? Has anyone implemented similar functionality?

**Comment by vqvu** [comment]
Does the download work with regular `request` instead of `retry-request`?

I tried downloading google.com and it worked (using Node v8.9.1), so I'm not sure what's going on. I had to set `Accept-Encoding` to get Google to send me zipped content, and I had to update the pipeline in `download` so that the downloads are actually done in parallel.

Here's the diff

```diff
diff --git a/test.orig.js b/test.js
index d5ab29b..6e3cac9 100644
--- a/test.orig.js
+++ b/test.js
@@ -9,10 +9,15 @@ function download (urls, path) {
   const output = fs.createWriteStream(tmpPath);
 
   const getUrl = url => {
+    console.log(`GET ${url}.`);
     return _((push, next) => {
       const options = {
         url,
-        retries: 10
+        retries: 10,
+        headers: {
+          'Accept': 'text/html',
+          'Accept-Encoding': 'gzip',
+        },
       };
       request(options)
         .on('response', resp => {
@@ -31,8 +36,9 @@ function download (urls, path) {
 
   return new Promise((resolve, reject) => {
     _(urls)
-      .flatMap(getUrl)
+      .map(getUrl)
       .parallel(5)
+      .sequence()
       .through(unzip)
       .stopOnError(err => error = err)
       .pipe(output)
@@ -48,3 +54,19 @@ function download (urls, path) {
       .on('error', reject);
   });
 }
+
+download([
+  'https://google.com?a=1',
+  'https://google.com?a=2',
+  'https://google.com?a=3',
+  'https://google.com?a=4',
+  'https://google.com?a=5',
+  'https://google.com?a=6',
+  'https://google.com?a=7',
+  'https://google.com?a=8',
+  'https://google.com?a=9',
+  'https://google.com?a=10',
+  'https://google.com?a=11',
+  'https://google.com?a=12',
+  'https://google.com?a=13',
+], 'output.txt').then(() => console.log('done.'), console.log);
```

**Comment by moose56** [comment]
Thank you for your investigation. I was not aware that I needed to add in `.sequence()` for parallel to work properly.

It works correctly when using `request`. I spent most of yesterday trying to make a reproducible example and work out what was happening. I can only reproduce the problem with live data and only in specific cases although I have not been able to work out what the particular difference is that causes the problem.

I can reproduce the issue when downloading a single file and the symptom is the program just exits while the data is being written to a file. Looking at the write stream evens i get a series of `drain` events called. Then execution exits and I cannot see why. `close` is never called on the write stream.

I have had to concede defeat and look for other options to achieve the retry functionality. It does appear that the problem is in `retry-request` not `highland` given that that is the bit that is changing in this example.

Thanks again for the reply.
